### PR TITLE
smtp/mime: configurable url scheme extraction v4

### PR DESCRIPTION
**Pull request by abungay**
Previous PR: #5698

Make sure these boxes are signed before submitting your Pull Request -- thank you.

- [x] I have read the contributing guide lines at https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Contributing
- [x] I have signed the Open Information Security Foundation contribution agreement at https://suricata-ids.org/about/contribution-agreement/
- [ ] I have updated the user guide (in doc/userguide/) to reflect the changes made (if applicable)

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/2054

Suricata-verify:
https://github.com/OISF/suricata-verify/pull/254

Describe changes:
- adds option 'extract-urls-schemes' to suricata.yaml to specify schemes to extract URLs instead of only 'http://' prefixed URL's getting parsed from mime bodies of SMTP traffic when using 'extract-urls: yes'
- adds option 'log-url-scheme' to suricata.yaml to specify whether to log the scheme portion of the extracted URLs, e.g. log the extra 'http://' prefix along with URLs

Changes since last PR:
- run clang-format.sh to fix c code format check failures

[PRScript](https://redmine.openinfosecfoundation.org/projects/suricata/wiki/PRscript) output (if applicable):
- Passes on local docker buildbot.

### CODE REVIEW

File: src/util-decode-mime.c [hunk 1]
```diff
@@ -1024,88 +1022,107 @@ static int FindUrlStrings(const uint8_t *line, uint32_t len,
...
+    uint8_t *fptr, *remptr, *tok = NULL, *tempUrl, *urlHost;
+    uint32_t tokLen = 0, i, tempUrlLen, urlHostLen;
+    uint8_t schemeStrLen = 0, flags = 0;
+    ConfNode *scheme = NULL;
+    char schemeStr[URL_SCHEME_SIZE + 3 + 1]; /* scheme + '://' + null terminator */
+    int r;
+
+    if (mdcfg != NULL && mdcfg->extract_urls_schemes != NULL) {
+        TAILQ_FOREACH (scheme, &mdcfg->extract_urls_schemes->head, next) {
+            r = snprintf(schemeStr, sizeof(schemeStr), "%s://", scheme->val);
```
Comment (victorjulien) [inline src/util-decode-mime.c:1035]: is this done at search time? If so, it feels like it belongs elsewhere. It should be a one time cost, not a per search cost.

File: src/util-decode-mime.c [hunk 2]
```diff
@@ -1024,88 +1022,107 @@ static int FindUrlStrings(const uint8_t *line, uint32_t len,
...
+                        flags = 0;
                     }
-                } else {
-                    SCFree(tempUrl);
                 }
-            }
+            } while (fptr != NULL);
         }
-    } while (fptr != NULL);
+    } else {
```
Comment (victorjulien) [inline src/util-decode-mime.c:1122]: I think if we test for this `else` condition first and just return `MIME_DEC_ERR_DATA` directly, we can avoid the above extra layer of indent?

File: src/util-decode-mime.c [hunk 3]
```diff
@@ -2793,9 +2810,17 @@ static int MimeDecParseLineTest02(void)
     uint32_t expected_count = 2;
     uint32_t line_count = 0;
 
+    ConfNode *url_schemes = ConfNodeNew();
+    ConfNode *scheme = ConfNodeNew();
+
+    url_schemes->is_seq = 1;
+    scheme->val = SCStrdup("http");
```
Comment (victorjulien) [inline src/util-decode-mime.c:2817]: Needs error check

File: src/util-decode-mime.h [hunk 4]
```diff
@@ -96,6 +97,9 @@ typedef struct MimeDecConfig {
     int decode_base64;  /**< Decode base64 bodies */
     int decode_quoted_printable;  /**< Decode quoted-printable bodies */
     int extract_urls;  /**< Extract and store URLs in data structure */
+    ConfNode *extract_urls_schemes; /**< List of schemes of which to
+                                         extract urls  */
+    int log_url_scheme;             /**< Log the scheme of extracted URLs */
```
Comment (victorjulien) [inline src/util-decode-mime.h:102]: can you switch this to `bool`? I see other existing code can be improved with that as well, if you're interested (in a separate commit)

File: src/app-layer-smtp.c [hunk 5]
```diff
@@ -284,6 +287,67 @@ static void SMTPConfigure(void) {
...
+            }
+
+            smtp_config.mime_config.extract_urls_schemes = extract_urls_schemes;
+        } else {
+            /* Add default extract url scheme 'http' since
+             * extract-urls-schemes wasn't found in the config */
+            ConfNode *seq_node;
+            ConfNode *scheme;
+
+            seq_node = ConfNodeNew();
```
Comment (victorjulien) [inline src/app-layer-smtp.c:319]: declare seq_node here

File: src/app-layer-smtp.c [hunk 6]
```diff
@@ -284,6 +287,67 @@ static void SMTPConfigure(void) {
...
+            if (unlikely(seq_node == NULL)) {
+                exit(EXIT_FAILURE);
```
Comment (victorjulien) [inline src/app-layer-smtp.c:321]: FatalError

File: src/app-layer-smtp.c [hunk 7]
```diff
@@ -284,6 +287,67 @@ static void SMTPConfigure(void) {
             smtp_config.mime_config.extract_urls = val;
         }
 
+        extract_urls_schemes = ConfNodeLookupChild(config, "extract-urls-schemes");
```
Comment (victorjulien) [inline src/app-layer-smtp.c:290]: I find it hard to parse this... what is happening here on a high level (please add comment)

**Review by victorjulien — CHANGES_REQUESTED** [review]

File: src/util-decode-mime.h (hunk 4 above)
Comment (abungay) [reply src/util-decode-mime.h:102]: Sure, will do!

File: src/util-decode-mime.c (hunk 1 above)
Comment (abungay) [reply src/util-decode-mime.c:1035]: Agreed, switched to configure time once instead of repeating every time a search is done.

**Comment by abungay** [comment]
Opened v5 PR - https://github.com/OISF/suricata/pull/5948
